core/corehttp/gateway_handler.go

```diff
@@ -38,6 +39,29 @@ type gatewayHandler struct {
 	api    coreiface.CoreAPI
 }
 
+// StatusResponseWriter enables us to override HTTP Status Code passed to
+// WriteHeader function inside of http.ServeContent.  Decision is based on
+// presence of HTTP Headers such as Location.
+type statusResponseWriter struct {
+	w http.ResponseWriter
```

Comment:
If you use:

```go
 type statusResponseWriter struct {
  http.ResponseWriter
}
```

You don't have to re-implement functions you're overriding: https://golang.org/doc/effective_go.html#embedding

Reply:
Initially I tried that, but was not able to make embedding work. 

When request for http://localhost:8080/ipfs/bafkreicysg23kiwv34eg2d7qweipxwosdo2py4ldv42nbauguluen5v6am/  is being processed I get a crash with a stack overflow, possibly due to prometheus client doing own wrapping which produces a circular call(?):

```
fatal error: stack overflow

runtime stack:
runtime.throw(0x1ad5596, 0xe)
        /usr/lib/go/src/runtime/panic.go:774 +0x72
runtime.newstack()
        /usr/lib/go/src/runtime/stack.go:1046 +0x6e9
runtime.morestack()
        /usr/lib/go/src/runtime/asm_amd64.s:449 +0x8f

goroutine 1169 [running]:
net/http.(*response).Header(0xc0005a8380, 0xc020f423e8)
        /usr/lib/go/src/net/http/server.go:1056 +0x8a fp=0xc020f42360 sp=0xc020f42358 pc=0xd6de7a
go.(*struct { *github.com/prometheus/client_golang/prometheus/promhttp.responseWriterDelegator; io.ReaderFrom; net/http.Hijacker; net/http.Flusher; net/http.CloseNotifier }).Header(0xc0003c4960, 0xc0007a2000)
        <autogenerated>:1 +0x3f fp=0xc020f42380 sp=0xc020f42360 pc=0x1704adf
go.(*struct { *github.com/prometheus/client_golang/prometheus/promhttp.responseWriterDelegator; io.ReaderFrom; net/http.Hijacker; net/http.Flusher; net/http.CloseNotifier }).Header(0xc0003c49b0, 0xc0003308d0)
        <autogenerated>:1 +0x3f fp=0xc020f423a0 sp=0xc020f42380 pc=0x1704adf
github.com/ipfs/go-ipfs/core/corehttp.(*statusResponseWriter).WriteHeader(0xc000303d90, 0x12d)
        src/github.com/ipfs/go-ipfs/core/corehttp/gateway_handler.go:52 +0x33 fp=0xc020f423d8 sp=0xc020f423a0 pc=0x1714763
github.com/ipfs/go-ipfs/core/corehttp.(*statusResponseWriter).WriteHeader(0xc000303d90, 0x12d)
        src/github.com/ipfs/go-ipfs/core/corehttp/gateway_handler.go:56 +0x7d fp=0xc020f42410 sp=0xc020f423d8 pc=0x17147ad
github.com/ipfs/go-ipfs/core/corehttp.(*statusResponseWriter).WriteHeader(0xc000303d90, 0x12d)
        src/github.com/ipfs/go-ipfs/core/corehttp/gateway_handler.go:56 +0x7d fp=0xc020f42448 sp=0xc020f42410 pc=0x17147ad
github.com/ipfs/go-ipfs/core/corehttp.(*statusResponseWriter).WriteHeader(0xc000303d90, 0x12d)
        src/github.com/ipfs/go-ipfs/core/corehttp/gateway_handler.go:56 +0x7d fp=0xc020f42480 sp=0xc020f42448 pc=0x17147ad
github.com/ipfs/go-ipfs/core/corehttp.(*statusResponseWriter).WriteHeader(0xc000303d90, 0x12d)
        src/github.com/ipfs/go-ipfs/core/corehttp/gateway_handler.go:56 +0x7d fp=0xc020f424b8 sp=0xc020f42480 pc=0x17147ad
github.com/ipfs/go-ipfs/core/corehttp.(*statusResponseWriter).WriteHeader(0xc000303d90, 0x12d)
[...]
```

Reply:
Ah, you were recursively calling `sw.WriteHeader`. You'd need to call `sw.ResponseWriter.WriteHeader`.

Reply:
Thanks! Rebased, this should be fixed in https://github.com/ipfs/go-ipfs/pull/6982/commits/9fe44b7fc29e3845a74283b6ec4a644b75fcd281